tests/cpp/tree/test_gpu_hist.cu

```diff
@@ -251,6 +251,9 @@ void TestHistogramIndexImpl() {
 
   // Build 2 matrices and build a histogram maker with that
   tree::GPUHistMakerSpecialised<GradientPairPrecise> hist_maker, hist_maker_ext;
+  auto pp_dmat = xgboost::CreateDMatrix(kNRows, kNCols, 0);
```

Comment:
I feel like `xgboost::CreateDMatrix` should just return a shared pointer. You don't have to do this in this PR.

Reply:
@RAMitchell Actually I did it in this PR previously then reverted.  The code change became quite scary.